## Q1. Explain the assumptions required to use ANOVA and provide examples of violations that could impact the validity of the results.

## Asnwer 

#### 1. Normality of sampling distribution of means
#### || The distribution of sample mean is normally distributed.
#### Example Violation : Suppose we conduct a one way anova test, if the data set is not following a bell cured shape then the assumptions can be violated.
##
#### 2. Absence of outliers
#### || Outliers score need to be removed from the data set.
#### Example Violation : It helps to reduce the type 1 error risk.
##
#### 3. Homogenity of variance
#### || Each one of the population has the same variance.
#### Example Violation : If the variance between the samples differ significantly then the final test result can be affected. and 
##
#### 4. Samples are independent and random to each other.
#### || Independent samples ensures that obervation within each group de=oes not affect EACH OTHER.

## 

##  Q2. What are the three types of ANOVA, and in what situations would each be used?

## Answer 

#### 1. One Way ANOVA 
#### || One Factor , =2 & >2 variables (Independent)
#
#### 2. Two Way ANOVA
#### || One Factor ,  at least 2 variable (Dependent)
#
#### 3. N>2 ANOVA
#### || 2 or more than 2 factors (each woth at least 2 levels), levels can be independent or dependent. 

## 

##  Q3. What is the partitioning of variance in ANOVA, and why is it important to understand this concept?

## Answer

### Partioning of Variance in ANOVA :
#### Between Group Variance :
#### || Let say we have a copare the heights of plants of different species, in B/W group variance we compare the average height of each species.
#### Within Group Variance :
#### || Now, We can also understand that there is also some variation in the height of plants of an individual specie. To Within group variance captures the variability in the heights of the plants of that particular species.
#
#### Why is it important ?
#### It is that important because, it hepls us to interpret and understand our result in a more good way. for example if the b/w variance is larger than within, it can easily be understood that one group mean is significantly larger than the other.
#### Where other tests, like z test , ttest, chi square tests, helps us to compare between 2 groups, on the other hand, ANOVA helps us to compare between 2 or more than 2 groups.

## 

## Q4. How would you calculate the total sum of squares (SST), explained sum of squares (SSE), and residual sum of squares (SSR) in a one-way ANOVA using Python?

## Answer 

#### To find these 3 in python, we'll need to create a linear regression line first.

In [19]:
from scipy.stats import f_oneway

## 

## Q5. In a two-way ANOVA, how would you calculate the main effects and interaction effects using Python?

## Answer 

#### to find this, we'll first require to import

In [21]:
from scipy.stats import f_oneway

#### within our code we'll create a linear regression line usinf f_oneway, and with this we can find our p_value and oneway stats.

#### If the pvalue is less than the significance value, we can reject the null hypothesis. else we'll fail to reject the null hypothesis.

## 

##  Q6. Suppose you conducted a one-way ANOVA and obtained an F-statistic of 5.23 and a p-value of 0.02. What can you conclude about the differences between the groups, and how would you interpret these results?

## Answer 

In [23]:
f_stats = 5.23
p_value = 0.02

#### 1. We can see here that p_value is lower than the significance value 0.05. Hence we can reject the null hypothesis.
#### 2. F_stats conclude that there is significant difference between means of each groups.

## 

##  Q7. In a repeated measures ANOVA, how would you handle missing data, and what are the potential consequences of using different methods to handle missing data?

## Answer 

#### Listwise-deletion
#### || Sometimes, we can see that we have so many missing values in the data. to overcome this, In listwise deletion menthod, we can delete the entire variable of missing values. It is not recommended to use this method unless there is significant missing values in the data.
### CONSEQUENCES :Loss of information , reduce the statistical power
#
#### Imputation Method
#### || In this method, we impute the estimated values from the observed data. We can replace the missing values with the mean or median of the variable.
### CONSEQUENCES : We assumes tha new value , Final result may have some degree of error.

## 

##  Q8. What are some common post-hoc tests used after ANOVA, and when would you use each one? Provide an example of a situation where a post-hoc test might be necessary.

## Answer 

#### After Comparing the means of the two different groups, to check which gruop is significantly different from the other one's, we use post-hoc tests.
#### Few of the post-hoc tests are mentioned below :
#### HSD ( Honest Significant Difference )Test : 
#### || It is very useful for pairwise COMPARISONS
#### || Suppose we want to measure the stamina of 3 different athlets, HSD Test will help us a lot to check which athlet's stamina differ from the another 2.
#### Example : Athelet A's stamina is different from B.
#
#### HOLM's Method
#### || Slightly more conservative than HSD Test.
#### Rather comparing the groups directly, it's more cautious. it ranks the Pvalue.
#### Example : If A's stamina is higher than B, then it must be higher than C.

##

## Q9. A researcher wants to compare the mean weight loss of three diets: A, B, and C. They collect data from 50 participants who were randomly assigned to one of the diets. Conduct a one-way ANOVA using Python to determine if there are any significant differences between the mean weight loss of the three diets. Report the F-statistic and p-value, and interpret the results.

In [2]:
n = 50

In [3]:
diet_A = [2.5, 3.0, 2.8, 2.3, 2.7, 2.5, 3.0, 2.8, 2.3, 2.7]
diet_B = [2.1, 2.4, 2.0, 2.5, 2.3, 1.8, 2.0, 1.9, 2.2, 1.7]
diet_C = [1.8, 2.0, 1.9, 2.2, 1.7, 2.1, 2.4, 2.0, 2.5, 2.3]

In [8]:
from scipy.stats import f_oneway

In [11]:
stats, pvalue = f_oneway(diet_A, diet_B, diet_C)

In [7]:
if pvalue<0.05 :
    print("We reject the Null Hypothesis")
else :
    print("We fail to reject the null hypothesis")

We reject the Null Hypothesis


In [15]:
stats

16.226970033296336

## As we know, more the stats value, more the difference between means of groups.

##

## Q10. A company wants to know if there are any significant differences in the average time it takes to complete a task using three different software programs: Program A, Program B, and Program C. They randomly assign 30 employees to one of the programs and record the time it takes each employee to complete the task. Conduct a two-way ANOVA using Python to determine if there are any main effects or interaction effects between the software programs and employee experience level (novice vs. experienced). Report the F-statistics and p-values, and interpret the results.

## Answer

In [36]:
from statsmodels.stats.anova import anova_lm
from statsmodels.formula.api import ols

In [37]:
import pandas as pd
data = {
    'program': ['A', 'B', 'C'] * 10,
    'experience': ['novice'] * 15 + ['experienced'] * 15,
    'completion_time': [10.23, 12.45, 11.78, 9.92, 14.01,13.20, 10.67, 9.34, 11.89, 14.56,12.10, 9.71, 13.80, 10.98, 14.32,11.05, 9.12, 13.67, 10.76, 14.88,12.33, 9.45, 11.34, 14.79, 12.67,9.87, 13.45, 11.56, 14.23, 12.89]
}
df = pd.DataFrame(data)

In [39]:
model = ols('completion_time ~ C(program) + C(experience) + C(program):C(experience)', data=df).fit()
anovaResults = anova_lm(model, type=2)

In [41]:
anovaResults

df     sum_sq   mean_sq         F    PR(>F)
C(program)                 2.0   9.536540  4.768270  1.460276  0.252074
C(experience)              1.0   0.320333  0.320333  0.098102  0.756826
C(program):C(experience)   2.0   4.953927  2.476963  0.758566  0.479239
Residual                  24.0  78.367720  3.265322       NaN       NaN

In [134]:
if anovaResults["PR(>F)"]["C(program):C(experience)"]<0.05:
    print("Significant change")
else:
    print("No significant change")

No significant change


In [ ]:
from stat import f_tx

## 

## Q11. An educational researcher is interested in whether a new teaching method improves student test scores. They randomly assign 100 students to either the control group (traditional teaching method) or the experimental group (new teaching method) and administer a test at the end of the semester. Conduct a two-sample t-test using Python to determine if there are any significant differences in test scores between the two groups. If the results are significant, follow up with a post-hoc test to determine which group(s) differ significantly from each other.

## Answer

In [123]:
import scipy.stats as stats
import pandas as pd
import numpy as np

In [124]:
data = pd.DataFrame({
    'group': ['C'] * 50 + ['E'] * 50,
'score': np.random.randint(45, 98, size=100)

})

In [125]:
TraditionalStyle = data[data["group"]=="C"]["score"]

In [126]:
ExpStyle = data[data["group"]=="E"]["score"]

In [127]:
stats, pvalue = stats.ttest_ind(TraditionalStyle, ExpStyle)

In [128]:
stats, pvalue

(0.7408134157192365, 0.460577936425705)

In [129]:
if pvalue<0.05:
    print("There is insignificant change in both the teaching methods")
else:
    print("there is no change in both the teaching methods")

there is no change in both the teaching methods


## POST_HOC 

In [130]:
from statsmodels.stats.multicomp import pairwise_tukeyhsd

In [132]:
posthoc = pairwise_tukeyhsd(data["score"], data["group"])

In [133]:
print(hsd_result.summary())

 Multiple Comparison of Means - Tukey HSD, FWER=0.05 
group1 group2 meandiff p-adj   lower    upper  reject
-----------------------------------------------------
     C      E    -6.52 0.0369 -12.6355 -0.4045   True
-----------------------------------------------------


#### We can see that the rejection is True, Means We reject the Null Hypothesis. There is significantly changes between both the groups.
#### The students in group E scored 6.52 points low than the students in Group C

## 

##  Q12. A researcher wants to know if there are any significant differences in the average daily sales of three retail stores: Store A, Store B, and Store C. They randomly select 30 days and record the sales for each store on those days. Conduct a repeated measures ANOVA using Python to determine if there are any significant differences in sales between the three stores. If the results are significant, follow up with a posthoc test to determine which store(s) differ significantly from each other.

## Answer 

In [74]:
data = {
    "sales" : np.random.randint(30, 80, size = 30),
    "store" : ["A", "B", "C"] *10
}

In [77]:
df = pd.DataFrame(data)

In [79]:
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm

In [81]:
model = ols("sales ~ store", data=df).fit()

In [83]:
results = anova_lm(model, type=2)

In [84]:
results

df       sum_sq     mean_sq         F    PR(>F)
store      2.0   685.066667  342.533333  1.408979  0.261794
Residual  27.0  6563.900000  243.107407       NaN       NaN

In [87]:
if results['PR(>F)']['store']<0.05:
    print("Insignificant Difference between the 3")
    posthoc_test = pairwise_tukeyhsd(data["sales"], data["store"])
else:
    print("There is no insignificant difference")

There is no insignificant difference
